spark: Set whether the write order of the table supports case-sensitive synchronization with sparkSQL

xiaotianzhang01 opened the pull request:
What changes were proposed in this pull request?
This PR is solved when sparkSQL configuration is case-insensitive, but must be case sensitive when `replaceSortOrder` is called.

Why are the changes needed?
In principle, whether to support case sensitivity should be configured and unified by the upper-layer engine

issue: https://github.com/apache/iceberg/issues/3523

Does this PR introduce any user-facing change?

xiaotianzhang01 commented:
cc: @rdblue @RussellSpitzer @danielcweeks @aokolnychyi

rdblue commented:
@xiaotianzhang01, can you add test cases that demonstrate the problem you're trying to solve? I asked for clarification on #3523, but it still isn't clear what's happening. I think a couple of test cases would show the problem.

xiaotianzhang01 commented:
okey, i'm going to add a couple of test cases

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/SortOrderBuilder.java:
```diff
@@ -106,4 +106,12 @@ default R desc(Term term) {
   R desc(Term term, NullOrder nullOrder);
+
+  /**
+   * whether the Projection should consider case sensitivity on column names or not.
```
Should be a bit more clear. 
"Whether to resolve column names in either a case-sensitive or insensitive way."?

RussellSpitzer commented:
I'm a little worried about adding this to the core api, should this be something that we figure out at the Spark side and pass in the correct identifiers from there? Like it seems like the SortOrderBuilder should be like the PartitionSpecBuilder and require exact matches. Then if we want to support case insensitivity we rely on the engine (or convert at the engine level) to get the correct identifiers.

xiaotianzhang01 commented:
@RussellSpitzer You're right, we should have a common util at the spark engine level for this, i updated the pr to put the convert int `Spark3Util`

kbendick commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java:
```diff
@@ -301,7 +302,9 @@ public static Term toIcebergTerm(Expression expr) {
           "Cannot convert transform with more than one column reference: %s", transform);
-      String colName = DOT.join(transform.references()[0].fieldNames());
+      List<String> fieldNames = Arrays.stream(transform.references()[0].fieldNames())
+              .map(Spark3Util::toIcebergField).collect(Collectors.toList());
```
Nit: This is over indented. The dot in `.map` should line up with the first double quote in the new line 304 (so 4 spaces indented past List).

xiaotianzhang01 replied:
done

kbendick commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java:
```diff
@@ -448,6 +451,19 @@ public static boolean extensionsEnabled(SparkSession spark) {
+
+  public static boolean isCaseSensitiveEnabled() {
+    return Boolean.parseBoolean(
+            SparkSession.active().conf().get("spark.sql.caseSensitive", "false"));
```
Nit: Same note about over-indentation. It should be aligned with the second `r` in return.

xiaotianzhang01 replied:
done

kbendick commented on spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSetWriteDistributionAndOrdering.java:
```diff
@@ -279,4 +281,40 @@ public void testSetWriteDistributedAndLocallyOrderedInverted() {
+    spark.conf().set("spark.sql.caseSensitive", "true");
+    AssertHelpers.assertThrows("Should reject invalid `write ordered` columns",
+            ValidationException.class, "Cannot find field 'CATEGORY' in struct",
+        () -> {
```
Nit: believe this lambda could fit on one line and there’s no need for the brackets.

kbendick reviewed (commented):
Thanks for working on this!

I left some notes on style related issues, mostly related to indentation.

I’m concerned whether or not this will work with the CachingCatalog. I don’t believe it respects case sensitivity being enabled or not. Wondering if others have the same concern. Once the caching catalog starts being used, will this change be somewhat inconsistent with the other behavior then?

xiaotianzhang01 commented:
@kbendick Thank you for your review and i have updated it.  I don't know much about `CachingCatalog` but i added some test cases for this. Are there any cases for other?

rdblue commented on spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSetWriteDistributionAndOrdering.java:
```diff
@@ -279,4 +281,38 @@ public void testSetWriteDistributedAndLocallyOrderedInverted() {
+        .withOrderId(1)
+        .asc("category", NullOrder.NULLS_FIRST)
+        .asc("id", NullOrder.NULLS_FIRST)
+        .build();
```
Please fix indentation throughout this PR.

rdblue commented on spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCaseSensitiveWithCacheEnabled.java:
```diff
@@ -0,0 +1,112 @@
+                    "type", "hive",
+                    "default-namespace", "default",
+                    "parquet-enabled", "true",
+                    "cache-enabled", "true"
```
Please fix indentation.

rdblue commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java:
```diff
@@ -448,6 +451,19 @@ public static boolean extensionsEnabled(SparkSession spark) {
 
+  public static String toIcebergField(String name) {
+    Preconditions.checkArgument(!name.isEmpty(), "Invalid column name: (empty)");
+    if (!isCaseSensitiveEnabled()) {
```
This isn't the right way to make sort order case insensitive. If the table was created with an upper case name, this would no longer work.

I think the right way to fix this problem is to set the `caseSensitive` config property in the `SortOrderBuilder`. That is currently defaulted to `true`. I think you need to expose a way to set that and call it appropriately.
